modules/standard/Allocators.chpl

```diff
@@ -0,0 +1,377 @@
+      }
+    }
+  }
+
+  @chpldoc.nodoc
+  inline proc chpl__deleteWithAllocator(const ref alloc: class, const obj) {
+    var p = c_ptrTo(obj);
+    if obj != nil then
+      obj!.deinit();
+    alloc.deallocate(p);
+  }
+  @chpldoc.nodoc
+  inline proc chpl__deleteWithAllocator(ref alloc: record, const obj) {
+    var p = c_ptrTo(obj);
+    if obj != nil then
+      obj!.deinit();
+    alloc.deallocate(p);
+  }
+
+  @chpldoc.nodoc
+  class _LockWrapper {
+    type lockType = ChapelLocks.chpl_LocalSpinlock;
+    var lockVar = new lockType();
+    inline proc lock() do lockVar.lock();
+    inline proc unlock() do lockVar.unlock();
+  }
+
+  /*
+    A simple allocator that acts as a wrapper around :proc:`CTypes.allocate`
+    and :proc:`CTypes.deallocate`.
+  */
+  class mallocWrapper: allocator {
+    /*
+      Allocate memory for ``n`` bytes.
+
+      :arg n: The number of bytes to allocate.
+      :returns: A ``c_ptr(void)`` to the allocated memory.
+    */
+    proc allocate(n: int): c_ptr(void) {
+      return CTypes.allocate(int(8), n.safeCast(c_size_t));
+    }
+    /*
+      Deallocate memory at ``p``.
+
+      :arg p: The pointer to the allocated memory.
+    */
+    proc deallocate(p: c_ptr(void)) {
+      CTypes.deallocate(p);
+    }
+  }
+
+
+  /*
+    The is a simple bump pointer allocator that is useful for allocating many
+    small objects.
+
+    This allocator makes a single memory allocation in a contiguous block,
+    calls to :proc:`~bumpPtrMemPool.allocate` will return the next available
+    address. This allocator is not thread-safe by default, but can be made
+    thread-safe by setting the :proc:`~bumpPtrMemPool.parSafe` parameter to
+    ``true``. By default, all memory allocated by :type:`bumpPtrMemPool` is aligned
+    to 16 bytes. This can be changed by setting the
+    :param:`~bumpPtrMemPool.alignment` parameter. An alignment of 0 will disable
+    alignment.
+
+    The :type:`bumpPtrMemPool` is initialized with a fixed size and allocates
+    memory from that fixed size. Attempting to allocate more memory than the
+    fixed size will result in an error. With checks enabled the program will
+    halt, otherwise memory errors will occur.
+
+    All memory allocated by :type:`bumpPtrMemPool` is deallocated when the
+    :type:`bumpPtrMemPool` instance is destroyed. The
+    :proc:`~bumpPtrMemPool.deallocate` method is a no-op.
+  */
+  class bumpPtrMemPool: allocator {
+    /*
+      Whether or not this allocator is thread-safe. if ``true``, this allocator
+      will use a lock to ensure that only one thread can allocate memory at a
+      time.
+    */
+    param parSafe: bool = false;
+    /*
+      The alignment of memory allocated by this allocator. If ``alignment`` is
+      greater than 0, the pointers returned by :proc:`~bumpPtrMemPool.allocate`
+      will be aligned to that alignment. ``alignment`` must be a power of 2.
+    */
+    param alignment: int = 16;
```

Comment:
I am doubtful that there is any benefit to making this `param`. Not a big deal if you want to keep it `param` for now, but maybe Future Work.